pilot/pkg/xds/debug.go

```diff
@@ -102,17 +102,21 @@ type AdsClients struct {
 
 // SyncStatus is the synchronization status between Pilot and a given Envoy
 type SyncStatus struct {
```

Comment:
We are only going to add more types (in fact, already added a new one). Maybe we need a more generic structure?

Reply:
Refactored this into a map from type to sync status. So it should be possible to easily support any XDS type. One concern is that this changes response structure of `/debug/syncz` endpoint. One consumer is `istioctl` and I fixed it. Do you know of any other consumers? Searched through the code and found another place where `SyncStatus` is used, but only version information is used and my change doesn't affect it anyways.

Reply:
Here is a sample output from `/debug/syncz` after refactoring
```json
    {
        "proxy": "productpage-v1-75665d56f8-l8czr.bookinfo",
        "istio_version": "1.10-dev",
        "statuses": {
            "CDS": {
                "nonce_sent": "8IiMpsy/OVg=54f33a8f-5fb1-4994-a9f9-524d9b4eb1e5",
                "nonce_acked": "8IiMpsy/OVg=54f33a8f-5fb1-4994-a9f9-524d9b4eb1e5",
                "last_sync_time": "2021-03-23T15:41:49.896512033Z"
            },
            "EDS": {
                "nonce_sent": "8IiMpsy/OVg=f4747e4f-3f4f-40c3-8a7e-23353cc0ca74",
                "nonce_acked": "8IiMpsy/OVg=f4747e4f-3f4f-40c3-8a7e-23353cc0ca74",
                "last_sync_time": "2021-03-23T15:41:49.896931264Z"
            },
            "LDS": {
                "nonce_sent": "8IiMpsy/OVg=7df6244c-d6c8-4a9d-bf48-3c37adc235c3",
                "nonce_acked": "8IiMpsy/OVg=7df6244c-d6c8-4a9d-bf48-3c37adc235c3",
                "last_sync_time": "2021-03-23T15:41:49.90279765Z"
            },
            "NDS": {
                "nonce_sent": "8IiMpsy/OVg=2c3ebc9a-cca1-436a-b5e2-a151f7d6e188",
                "nonce_acked": "8IiMpsy/OVg=2c3ebc9a-cca1-436a-b5e2-a151f7d6e188",
                "last_sync_time": "2021-03-23T15:41:49.903217834Z"
            },
            "PCDS": {
                "nonce_sent": "8IiMpsy/OVg=497d8aca-b811-4f56-b2f1-4cb2d5599011",
                "nonce_acked": "8IiMpsy/OVg=497d8aca-b811-4f56-b2f1-4cb2d5599011",
                "last_sync_time": "2021-03-23T15:41:49.903456225Z"
            },
            "RDS": {
                "nonce_sent": "8IiMpsy/OVg=cb86a20f-5e70-4832-8608-2b3fe50fe8c1",
                "nonce_acked": "8IiMpsy/OVg=cb86a20f-5e70-4832-8608-2b3fe50fe8c1",
                "last_sync_time": "2021-03-23T15:41:49.904829705Z"
            }
        }
    },
```
When a new `typeURL` appears in the list of `watchedResources` for the proxy it **automatically appears here**. It is up to the clients to decide whether to use this information or not.